File: LeetCode/iter1/c++/543.diameter-of-binary-tree.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

struct TreeNode {
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	int diameterOfBinaryTree(TreeNode* root) {
		if (root == NULL) return 0;

		diameter = 0;
		getTreeDepth(root);

		return diameter - 1;
	}
private:
	int diameter;

	int getTreeDepth(TreeNode* root) {
		if (root == NULL) return 0;

		int LDepth = getTreeDepth(root->left);
		int RDepth = getTreeDepth(root->right);

		diameter = max(diameter, LDepth + RDepth + 1);

		return max(LDepth, RDepth) + 1;
	}
};

TreeNode* testCase1() {
	TreeNode* node1 = new TreeNode(1);
	TreeNode* node2 = new TreeNode(2);
	TreeNode* node3 = new TreeNode(3);
	TreeNode* node4 = new TreeNode(4);
	TreeNode* node5 = new TreeNode(5);

	node2->left = node4;
	node2->right = node5;

	node1->left = node2;
	node1->right = node5;

	return node1;
}

TreeNode* testCase2() {
	//	1
	//2     3
	//4 5
	//6   7
	//8    9
	//      10
	TreeNode* node1 = new TreeNode(1);
	TreeNode* node2 = new TreeNode(2);
	TreeNode* node3 = new TreeNode(3);
	TreeNode* node4 = new TreeNode(4);
	TreeNode* node5 = new TreeNode(5);
	TreeNode* node6 = new TreeNode(6);
	TreeNode* node7 = new TreeNode(7);
	TreeNode* node8 = new TreeNode(8);
	TreeNode* node9 = new TreeNode(9);
	TreeNode* node10 = new TreeNode(10);

	node9->left = node10;
	node7->right = node9;
	node5->right = node7;

	node6->right = node8;
	node4->left = node6;

	node2->left = node4;
	node2->right = node5;

	node1->left = node2;
	node1->right = node3;

	return node1;
}

int main()
{
	Solution solution;
	cout << solution.diameterOfBinaryTree(testCase1()) << endl;
	cout << solution.diameterOfBinaryTree(testCase2()) << endl;

	return 0;
}

File: LeetCode/iter1/java/543.diameter-of-binary-tree.java
package org.ironman.base.playground.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Solution {

    public static void main(String[] args) {
        System.out.println(diameterOfBinaryTree((testCase1())));
        System.out.println(diameterOfBinaryTree((testCase2())));
        System.out.println(diameterOfBinaryTree((testCase3())));
    }

    private static int diameter = 0;

    public static int diameterOfBinaryTree(TreeNode root) {
        if (root == null) {
            return 0;
        }

        diameter = 0;
        getTreeDepth(root);

        return diameter - 1;
    }

    private static int getTreeDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int LDepth = getTreeDepth(root.left);
        int RDepth = getTreeDepth(root.right);

        diameter = Math.max(diameter, LDepth + RDepth + 1);

        return Math.max(LDepth, RDepth) + 1;
    }

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    //	1
    //2     3
    //4 5
    private static TreeNode testCase1() {
        List<TreeNode> lst = new ArrayList<>();
        for (int i = 0; i <= 5; i++) {
            lst.add(new TreeNode(i));
        }

        lst.get(2).left = lst.get(4);
        lst.get(2).right = lst.get(5);

        lst.get(1).left = lst.get(2);
        lst.get(1).right = lst.get(3);

        return lst.get(1);
    }

    //	1
    //2     3
    //4 5
    //6   7
    //8    9
    //      10
    private static TreeNode testCase2() {
        List<TreeNode> lst = new ArrayList<>();
        for (int i = 0; i <= 10; i++) {
            lst.add(new TreeNode(i));
        }

        lst.get(9).right = lst.get(10);
        lst.get(7).left = lst.get(9);
        lst.get(5).right = lst.get(7);

        lst.get(6).left = lst.get(8);
        lst.get(4).right = lst.get(6);

        lst.get(2).left = lst.get(4);
        lst.get(2).right = lst.get(5);

        lst.get(1).left = lst.get(2);
        lst.get(1).right = lst.get(3);

        return lst.get(1);
    }

    private static TreeNode testCase3() {
        return new TreeNode(1);
    }
}